File: src/KM.Api/RaidDens/RaidDensContracts.cs
// SPDX-License-Identifier: GPL-3.0-only

using KM.Api.Diagnostics;
using KM.Api.Editing;
using KM.Api.Projects;
using System.Text.Json.Serialization;

namespace KM.Api.RaidDens;

public sealed record LoadRaidDensRequest(ProjectPathsDto Paths);
public sealed record StageRaidDensRequest(ProjectPathsDto Paths, [property: JsonRequired] bool Disabled, EditSessionDto? Session);
public sealed record RaidDensWorkflowDto(bool CanEdit, bool? Disabled, string SourceLayer,
    ProjectGameDto? DetectedGame, IReadOnlyList<ApiDiagnostic> Diagnostics);
public sealed record LoadRaidDensResponse(RaidDensWorkflowDto Workflow);
public sealed record StageRaidDensResponse(RaidDensWorkflowDto Workflow, EditSessionDto Session,
    IReadOnlyList<ApiDiagnostic> Diagnostics);

File: apps/desktop/src/bridge/raidDensContracts.ts
/* SPDX-License-Identifier: GPL-3.0-only */
import { z } from 'zod';
import { apiDiagnosticSchema, editSessionSchema, projectGameSchema, projectPathsSchema } from './contracts';

export const raidDensWorkflowSchema = z.strictObject({
  canEdit: z.boolean(), disabled: z.boolean().nullable(), sourceLayer: z.string(),
  detectedGame: projectGameSchema.nullable(), diagnostics: z.array(apiDiagnosticSchema)
});
export const loadRaidDensRequestSchema = z.strictObject({ paths: projectPathsSchema });
export const stageRaidDensRequestSchema = z.strictObject({ paths: projectPathsSchema,
  disabled: z.boolean(), session: editSessionSchema.nullable() });
export const loadRaidDensResponseSchema = z.strictObject({ workflow: raidDensWorkflowSchema });
export const stageRaidDensResponseSchema = z.strictObject({ workflow: raidDensWorkflowSchema,
  session: editSessionSchema, diagnostics: z.array(apiDiagnosticSchema) });
export type RaidDensWorkflow = z.infer<typeof raidDensWorkflowSchema>;
export type LoadRaidDensRequest = z.infer<typeof loadRaidDensRequestSchema>;
export type LoadRaidDensResponse = z.infer<typeof loadRaidDensResponseSchema>;
export type StageRaidDensRequest = z.infer<typeof stageRaidDensRequestSchema>;
export type StageRaidDensResponse = z.infer<typeof stageRaidDensResponseSchema>;
